src/cudamatrix/cu-kernels.cu

```diff
@@ -2599,6 +2599,60 @@ static void _diff_softmax(Real* x, const MatrixDim dim, const Real* value,
   }
 }
 
+// Differentiate backward through the log softmax function.
+// "value" is the log softmax output. Does, for each row i,
+// x(i) =  diff(i) - sum(diff(i)) .* exp(value(i))
+// ???(i) is row-vector.
+// CUDA thread layout: one thread block (256 threads) per matrix-row.
+template<typename Real>
+__global__
+static void _diff_log_softmax(Real* x, const MatrixDim dim, const Real* value,
```

Comment:
We generally prefer the outputs of functions to be the final argument, as it's more consistent with the Google style... OK if this function was pre-existing.
Perhaps more suitable names would be:
 x -> input_deriv, or x_deriv
 value -> output, or y
 diff -> output_deriv, or y_deriv